### PR TITLE
GitCommitBear: Add imperative check in shortlog

### PR DESCRIPTION
**Pull request by tushar-rishav**
This commit adds `check_imperative` method in `GitCommitBear`.
The method checks if a shortlog is in imperative mood or not.
If the shortlog isn't in imperative mood then it'd throw an error
message including the invalid words that are the cause for being non
imperative so that user could get some hint and fix his commit.

Fixes https://github.com/coala-analyzer/coala-bears/issues/243

### CODE REVIEW

File: bears/vcs/git/GitCommitBear.py [hunk 1]
```diff
@@ -109,6 +110,36 @@ def check_shortlog(self,
...
+            yield Result(self,
+                         "Shortlog of HEAD commit isn't imperative mood,"
+                         "bad words are '{0}'".format(bad_words))
+
+    def check_imperative(self, paragraph):
+        """
+        Check the given sentence/s for Imperatives
+
+        :param paragraph: The input paragraph to be tested
+        :returns: 'Pass' or 'Fail'
```
Comment (SanketDG) [inline bears/vcs/git/GitCommitBear.py]: IIUC it returns the words?

Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Also, formatting (and period at the end)

File: requirements.txt [hunk 2]
```diff
@@ -23,3 +23,4 @@ pydocstyle~=1.0.0
 cmakelint~=1.3.4
 requests_mock~=0.7.0
 vim-vint~=0.3.6
+nltk
```
Comment (SanketDG) [inline requirements.txt]: current version should be provided like the others

File: bears/vcs/git/GitCommitBear.py (hunk 1 above)
Comment (SanketDG) [inline bears/vcs/git/GitCommitBear.py]: also provide doctests if you can.

File: requirements.txt (hunk 2 above)
Comment (AbdealiJK) [inline requirements.txt]: no version ?

File: bears/vcs/git/GitCommitBear.py [hunk 3]
```diff
@@ -109,6 +110,36 @@ def check_shortlog(self,
...
+        f_sentence = nltk.sent_tokenize(paragraph)[0]
+        words = nltk.word_tokenize(f_sentence)
+        flaws = []
+        # VBZ : Verb, 3rd person singular present, like 'adds', 'writes' etc
+        # VBD : Verb, Past tense , like 'added', 'wrote' etc
+        # VBG : Verb, Present participle, like 'adding', 'writing'
+        for word, tag in nltk.pos_tag(['I'] + words)[1:]:
+            if word.endswith('ing'):    # Handle special case for VBG
+                flaws.append((word, tag))
+            elif(tag.startswith('VBZ') or
```
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Can't you just put "or" in the earlier if and remove this elif ? (They both do the same thing if true)

File: bears/vcs/git/GitCommitBear.py [hunk 4]
```diff
@@ -109,6 +110,36 @@ def check_shortlog(self,
...
+        :param paragraph: The input paragraph to be tested
+        :returns: 'Pass' or 'Fail'
+        """
+        f_sentence = nltk.sent_tokenize(paragraph)[0]
+        words = nltk.word_tokenize(f_sentence)
+        flaws = []
+        # VBZ : Verb, 3rd person singular present, like 'adds', 'writes' etc
+        # VBD : Verb, Past tense , like 'added', 'wrote' etc
+        # VBG : Verb, Present participle, like 'adding', 'writing'
+        for word, tag in nltk.pos_tag(['I'] + words)[1:]:
```
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Hmmmmm. This is interesting. Have you read this in some SO post or something (about adding I in front of the words) ?

File: tests/vcs/git/GitCommitBearTest.py [hunk 5]
```diff
@@ -139,6 +139,17 @@ def test_shortlog_checks_shortlog_trailing_period(self):
...
+    def test_shortlog_checks_imperative(self):
+        self.git_commit("tag: Add shortlog in imperative")
+        self.assertNotIn("Shortlog of HEAD commit isn't imperative mood",
+                         ''.join(self.run_uut()))
+        self.git_commit("Added invalid shortlog")
+        self.assertIn("Shortlog of HEAD commit isn't imperative mood",
+                      ''.join(self.run_uut()))
+        self.git_commit("Adding another invalid shortlog")
+        self.assertIn("Shortlog of HEAD commit isn't imperative mood",
+                      ''.join(self.run_uut()))
```
Comment (AbdealiJK) [inline tests/vcs/git/GitCommitBearTest.py]: What if I have "Add the maximizing function" ? The  "ing" would be caught ?

File: bears/vcs/git/GitCommitBear.py (hunk 1 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Damn. Ah it returns a list. Yeah typo. :+1:

File: requirements.txt (hunk 2 above)
Comment (tushar-rishav) [inline requirements.txt]: I am using the core API and It's very rare that would change. Is it necessary to provide the version here?

File: bears/vcs/git/GitCommitBear.py (hunk 3 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Ah right. Sure :+1:

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Yeah. So I explained about the current *drawbacks* in nltk module. nltk isn't smart in handling verbs. Sometimes it'd report verbs as noun. I feel bit lazy to type again. See [here](https://github.com/coala-analyzer/coala-bears/issues/243#issuecomment-204484597) please  :D

File: tests/vcs/git/GitCommitBearTest.py (hunk 5 above)
Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: Sadly. Yes. I had this corner case caught and explained [here](https://github.com/coala-analyzer/coala-bears/issues/243#issuecomment-204495739). Actually, it's bit difficult to consider all the corner cases. I guess we can improve by considering the obvious ones.
Any thoughts?

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Looks good, but rigorous testing needed as it does indeed seem like a hack.

Also, won't the capital "A" in add affect NLTK ?

File: requirements.txt (hunk 2 above)
Comment (tushar-rishav) [inline requirements.txt]: IIUC not providing a version would fetch the latest stable version. Anything wrong with that?

Comment (AbdealiJK) [inline requirements.txt]: In case nltk makes a shiny new version next year, coala users who are using coala 0.7.0 (which expects nltk's version X) would be installing the X+1 version of nltk from pip.
Hence it'd never work. So, yeah - version is necessary.

Note that very rare change is not quite true ... It doesnt harm to use the `~=` syntax. so why not use it ?

File: tests/vcs/git/GitCommitBearTest.py (hunk 5 above)
Comment (AbdealiJK) [inline tests/vcs/git/GitCommitBearTest.py]: Can't we fid the word which is a "verb" ?
Because "maximizing function" is the object in that sentence right (and "maximizing" is adjective? (not sure)) ? So we should be checking only the word which is POS tagged as verb

File: requirements.txt (hunk 2 above)
Comment (tushar-rishav) [inline requirements.txt]: Ah sure. :+1:

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Um still It's a verb right? `Add some messages` is fine with nltk. It's imperative.

Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Hm, i see. I assumed it may consider capital beginning to be a proper noun maybe ... alright

File: tests/vcs/git/GitCommitBearTest.py (hunk 5 above)
Comment (AbdealiJK) [inline tests/vcs/git/GitCommitBearTest.py]: Hey, you know what would be a very simple test for this ? Just do `git log --oneline > /someplace/file.txt` and make a quick for loop checking this ?
If you could check if this works out - I'd be much more confident. (There may be a few wrong things ... Which we can manually identify as mistakes on our side)

Preferably make a paste bin and ass a link in the commit message of this commit ?

Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: Ah sadly nltk isn't that smart. It would classify "maximizing" in Present participle category. We can solve this problem by restricting users to follow certain grammar structure. Something like 
Verb | <anything>. So in that case we would test the first word only. Still it won't be perfect. For eg.
`tag: Add. This added stuff`. 
Clearly this ain't imperative. :D

Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: So IMO we can find many cases that'd fail the check. I am trying to cover the most common structures. Hm considering the dilemma the problem requires more thoughts maybe? Maybe an overkill? :P

Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: So if we check for word having POS tagged as Verb then it would certainly fail cases like `Added, Adding ` etc. Because they all are verb.

Comment (AbdealiJK) [inline tests/vcs/git/GitCommitBearTest.py]: I think maybe we want to either document or find solutions to the things it
cannot do at the very least

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: And btw it's indeed a hack. Sadly that's because nltk isn't good at distinguishing between noun and verb in *some* cases.

File: .ci/deps.osx.sh [hunk 6]
```diff
@@ -37,6 +37,9 @@ pip install -U pip
 pip install -r test-requirements.txt
 pip install -r requirements.txt
 
+# Downloading nltk data that's required for nltk to run 
+python -m nltk.downloader all
```
Comment (tushar-rishav) [inline .ci/deps.osx.sh]: Should probably cache it.

File: .ci/deps.sh [hunk 7]
```diff
@@ -57,6 +57,9 @@ done
 
 pip install -r docs-requirements.txt
 
+# Downloading nltk data that's required for nltk to run 
+python -m nltk.downloader all
```
Comment (tushar-rishav) [inline .ci/deps.sh]: Same as above. Cache it

File: .ci/deps.nltk.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
+set -e
+set -x
+
+
+NLTK_DATA_PATH=~/nltk_data
+
+function download_ntlk_data {
```
Comment (tushar-rishav) [inline .ci/deps.nltk.sh]: Fix the typo in next push.

File: bears/vcs/git/GitCommitBear.py [hunk 9]
```diff
@@ -109,6 +110,35 @@ def check_shortlog(self,
                 yield Result(
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
+        colon_pos = shortlog.find(':')
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: would be good if you introduce a `\n` before this one, as this is new functionality, makes it easier to read^^

Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: And btw I want an option for enabling/disabling this "imperative" check, sometimes there could be false-positives (seems you have some) and then one would like maybe to disable this feature. By default it can be enabled I think^^

File: bears/vcs/git/GitCommitBear.py [hunk 10]
```diff
@@ -109,6 +110,35 @@ def check_shortlog(self,
...
+        shortlog = shortlog[colon_pos:] if colon_pos != -1 else shortlog
+        has_flaws = self.check_imperative(shortlog)
+        if has_flaws:
+            bad_words = ', '.join([words[0] for words in has_flaws])
+            yield Result(self,
+                         "Shortlog of HEAD commit isn't imperative mood,"
+                         "bad words are '{0}'".format(bad_words))
+
+    def check_imperative(self, paragraph):
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: can be made a `@staticmethod`

File: bears/vcs/git/GitCommitBear.py [hunk 11]
```diff
@@ -109,6 +110,35 @@ def check_shortlog(self,
...
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
+        colon_pos = shortlog.find(':')
+        shortlog = shortlog[colon_pos:] if colon_pos != -1 else shortlog
+        has_flaws = self.check_imperative(shortlog)
+        if has_flaws:
+            bad_words = ', '.join([words[0] for words in has_flaws])
+            yield Result(self,
+                         "Shortlog of HEAD commit isn't imperative mood,"
+                         "bad words are '{0}'".format(bad_words))
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: `{0}` -> `{}`

File: tests/vcs/git/GitCommitBearTest.py [hunk 12]
```diff
@@ -109,17 +109,17 @@ def test_pure_oneliner_message(self, patch):
         self.assertTrue(self.msg_queue.empty())
 
     def test_shortlog_checks_length(self):
-        self.git_commit("Commits messages that nearly exceeds default limit")
+        self.git_commit("Commit messages that nearly exceed default limit")
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: please don't change this one, though it's wrong english the number of chars in it is on purpose, as (like it tries to state) uses up 50 chars which nearly exceeds the limit.
If you want put a comment above that^^

Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: Or change the commit message but keep its length

File: tests/vcs/git/GitCommitBearTest.py [hunk 13]
```diff
@@ -109,17 +109,17 @@ def test_pure_oneliner_message(self, patch):
...
 
         self.assertEqual(self.run_uut(), [])
         self.assertTrue(self.msg_queue.empty())
 
         self.assertEqual(self.run_uut(shortlog_length=17),
-                         ["Shortlog of HEAD commit is 33 character(s) "
-                          "longer than the limit (50 > 17)."])
+                         ["Shortlog of HEAD commit is 31 character(s) "
+                          "longer than the limit (48 > 17)."])
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: see above

File: tests/vcs/git/GitCommitBearTest.py [hunk 14]
```diff
@@ -109,17 +109,17 @@ def test_pure_oneliner_message(self, patch):
...
         self.assertTrue(self.msg_queue.empty())
 
-        self.git_commit("A shortlog that is too long is not good for history")
+        self.git_commit("Very long shortlog never good for a project history")
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: okay... but please in a separate commit

File: tests/vcs/git/GitCommitBearTest.py [hunk 15]
```diff
@@ -203,7 +214,7 @@ def test_body_checks(self):
 
     def test_different_path(self):
         no_git_dir = mkdtemp()
-        self.git_commit("A shortlog that is too long is not good for history")
+        self.git_commit("Very long shortlog never good for a project history")
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: also: goes into the same separate commit like the one above

File: bears/vcs/git/GitCommitBear.py [hunk 16]
```diff
@@ -109,6 +110,35 @@ def check_shortlog(self,
...
+        f_sentence = nltk.sent_tokenize(paragraph)[0]
+        words = nltk.word_tokenize(f_sentence)
+        flaws = []
+        # VBZ : Verb, 3rd person singular present, like 'adds', 'writes' etc
+        # VBD : Verb, Past tense , like 'added', 'wrote' etc
+        # VBG : Verb, Present participle, like 'adding', 'writing'
+        for word, tag in nltk.pos_tag(['I'] + words)[1:]:
+            if(tag.startswith('VBZ') or
+               tag.startswith('VBD') or
+               tag.startswith('VBG')or
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: missing space between `)` and `or` :)

File: bears/vcs/git/GitCommitBear.py [hunk 17]
```diff
@@ -109,6 +110,35 @@ def check_shortlog(self,
...
+        :returns: A list of invalid words.
+        """
+        f_sentence = nltk.sent_tokenize(paragraph)[0]
+        words = nltk.word_tokenize(f_sentence)
+        flaws = []
+        # VBZ : Verb, 3rd person singular present, like 'adds', 'writes' etc
+        # VBD : Verb, Past tense , like 'added', 'wrote' etc
+        # VBG : Verb, Present participle, like 'adding', 'writing'
+        for word, tag in nltk.pos_tag(['I'] + words)[1:]:
+            if(tag.startswith('VBZ') or
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: wondering why PEP allows this... we usually leave a space between the `if` and the expression, but as PEP8 allows it, nvm^^

File: bears/vcs/git/GitCommitBear.py (hunk 16 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: _/\_ :D

File: tests/vcs/git/GitCommitBearTest.py (hunk 14 above)
Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: Separate commit? It's a part of test. If I make these changes in separate commit it'd possibly fail?

File: .ci/deps.nltk.sh [hunk 18]
```diff
@@ -0,0 +1,17 @@
+set -e
+set -x
+
+
+NLTK_DATA_PATH=~/nltk_data
```
Comment (Makman2) [inline .ci/deps.nltk.sh]: don't you need to register this path to the cache?

Comment (tushar-rishav) [inline .ci/deps.nltk.sh]: I guess I've done this already. Please check `cache_directories` in
circle.yml once.
ᐧ

File: bears/vcs/git/GitCommitBear.py [hunk 19]
```diff
@@ -109,6 +113,37 @@ def check_shortlog(self,
                 yield Result(
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
+        if shortlog_imperative_check:
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: please put a newline before this line :)

File: .ci/deps.nltk.sh (hunk 18 above)
Comment (Makman2) [inline .ci/deps.nltk.sh]: never mind...

File: bears/vcs/git/GitCommitBear.py (hunk 19 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Sure. I'll push the updates once you are done reviewing. :D
ᐧ

File: bears/vcs/git/GitCommitBear.py [hunk 20]
```diff
@@ -109,6 +113,37 @@ def check_shortlog(self,
...
+                             "Shortlog of HEAD commit isn't imperative mood,"
+                             "bad words are '{}'".format(bad_words))
+
+    @staticmethod
+    def check_imperative(paragraph):
+        """
+        Check the given sentence/s for Imperatives.
+
+        :param paragraph: The input paragraph to be tested.
+        :returns: A list of invalid words.
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: oh please indent the `:returns:` docs to the `paragraph` one. Or use this (new and coming) style:
```
:param paragraph:
    docs...
:returns:
    stuff...
```

File: bears/vcs/git/GitCommitBear.py [hunk 21]
```diff
@@ -109,6 +113,37 @@ def check_shortlog(self,
                 yield Result(
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
+        if shortlog_imperative_check:
+            colon_pos = shortlog.find(':')
+            shortlog = shortlog[colon_pos:] if colon_pos != -1 else shortlog
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: `shortlog[colon_pos:]` would also extract the colon. Shouldn't it be `shortlog[colon_pos + 1:]`?

File: bears/vcs/git/GitCommitBear.py (hunk 20 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Whoa :P

File: bears/vcs/git/GitCommitBear.py [hunk 22]
```diff
@@ -110,6 +114,38 @@ def check_shortlog(self,
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
 
+        if shortlog_imperative_check:
+            colon_pos = shortlog.find(':')
+            shortlog = shortlog[colon_pos:] if colon_pos != -1 else shortlog
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: `shortlog[colon_pos:]` extracts the colon too, shouldn't it be `shortlog[colon_pos + 1:]`?

File: bears/vcs/git/GitCommitBear.py (hunk 20 above)
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: Please be also more specific, the list contains tuples with `word, tag` (I don't know what the tags are...)

File: bears/vcs/git/GitCommitBear.py [hunk 23]
```diff
@@ -110,6 +114,38 @@ def check_shortlog(self,
...
+    @staticmethod
+    def check_imperative(paragraph):
+        """
+        Check the given sentence/s for Imperatives.
+
+        :param paragraph: The input paragraph to be tested.
+        :returns: A list of invalid words.
+        """
+        f_sentence = nltk.sent_tokenize(paragraph)[0]
+        words = nltk.word_tokenize(f_sentence)
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: optional: as you don't use `f_sentence` any more you could maybe inline it into `words` variable^^

File: bears/vcs/git/GitCommitBear.py [hunk 24]
```diff
@@ -110,6 +114,38 @@ def check_shortlog(self,
                     self,
                     "Shortlog of HEAD commit does not match given regex.")
 
+        if shortlog_imperative_check:
+            colon_pos = shortlog.find(':')
+            shortlog = shortlog[colon_pos:] if colon_pos != -1 else shortlog
+            has_flaws = self.check_imperative(shortlog)
+            if has_flaws:
+                bad_words = ', '.join([words[0] for words in has_flaws])
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: if you want it faster:
you can use `map` and `operator.itemgetter`, `map` tries to vectorize the calculations then.
```
>>> from operator import itemgetter
>>> rows = [(1, 2), (3, 4), (5, 6)]
>>> map(itemgetter(1), rows)
[2, 4, 6]
```

Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: But only if you want :)

File: bears/vcs/git/GitCommitBear.py (hunk 22 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Thankyou didn't notice that. Awesome. :+1: :D
ᐧ

File: tests/vcs/git/GitCommitBearTest.py [hunk 25]
```diff
@@ -139,6 +139,20 @@ def test_shortlog_checks_shortlog_trailing_period(self):
         self.assertEqual(self.run_uut(shortlog_trailing_period=False), [])
         self.assertEqual(self.run_uut(shortlog_trailing_period=None), [])
 
+    def test_shortlog_checks_imperative(self):
+        self.git_commit("tag: Add shortlog in imperative")
+        self.assertNotIn("Shortlog of HEAD commit isn't imperative mood",
+                         ''.join(self.run_uut()))
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: is there a reason why you don't use lists for comparison like the other tests above? In fact they are slightly better, as it could be that `run_uut()` spits out more than one item that yield concatenated your string (highly unlikely, but anyway). Also it's more complicated to read imo^^

File: bears/vcs/git/GitCommitBear.py (hunk 24 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Is it fast? Wasn't aware of that. I'll look into that. Thanks.
ᐧ

File: bears/vcs/git/GitCommitBear.py [hunk 26]
```diff
@@ -110,6 +114,38 @@ def check_shortlog(self,
...
+                yield Result(self,
+                             "Shortlog of HEAD commit isn't imperative mood,"
+                             "bad words are '{}'".format(bad_words))
+
+    @staticmethod
+    def check_imperative(paragraph):
+        """
+        Check the given sentence/s for Imperatives.
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: You say sentence***s***. Do you accept sequences of strings?

Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: You meant something like `["whoa", "added some stuff"]` ? Yes that'd work
too
ᐧ

**Comment by tushar-rishav** [comment]
Add a check for `nltk_data`. If it's not present then just put up a helpful message asking users to install it before proceeding ahead. @tushar-rishav

File: tests/vcs/git/GitCommitBearTest.py [hunk 27]
```diff
@@ -119,7 +119,7 @@ def test_shortlog_checks_length(self):
                           "longer than the limit (50 > 17)."])
         self.assertTrue(self.msg_queue.empty())
 
-        self.git_commit("A shortlog that is too long is not good for history")
+        self.git_commit("Very long shortlog never good for a project history")
```
Comment (Adrianzatreanu) [inline tests/vcs/git/GitCommitBearTest.py]: this sentence doesn't have a verb.

File: tests/vcs/git/GitCommitBearTest.py [hunk 28]
```diff
@@ -203,7 +217,7 @@ def test_body_checks(self):
 
     def test_different_path(self):
         no_git_dir = mkdtemp()
-        self.git_commit("A shortlog that is too long is not good for history")
+        self.git_commit("Very long shortlog never good for a project history")
```
Comment (Adrianzatreanu) [inline tests/vcs/git/GitCommitBearTest.py]: no verb here either.

Comment (tushar-rishav) [inline tests/vcs/git/GitCommitBearTest.py]: IIUC "long" is actually a verb. A verb may not always be something that
shows action. It could also be a description.  :P

Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: @Adrianzatreanu come on, it's a test :)

Comment (sils1297) [inline tests/vcs/git/GitCommitBearTest.py]: @Makman2 he's on the right track to becoming a good reviewer don't you see :)

Comment (sils1297) [inline tests/vcs/git/GitCommitBearTest.py]: /me proudly pats onto @Adrianzatreanu s shoulder

File: bears/vcs/git/GitCommitBear.py [hunk 29]
```diff
@@ -110,6 +114,38 @@ def check_shortlog(self,
...
+
+        :param paragraph: The input paragraph to be tested.
+        :returns: A list of invalid words.
+        """
```
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: handling for non-present nltk data still missing also some of the other comments not yet addressed

File: tests/vcs/git/GitCommitBearTest.py (hunk 5 above)
Comment (sils1297) [inline tests/vcs/git/GitCommitBearTest.py]: hm false positives are an issue and IMO worse than false negatives, @tushar-rishav can you write some documentation about all this into the doc comment? Then it's visible to the user too.

File: bears/vcs/git/GitCommitBear.py (hunk 29 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: Lol. I am yet to push. :P

Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: marking wip then so no more review time is wasted

**Comment by tushar-rishav** [comment]
@sils1297  @Makman2 Need your blessings in this PR :P

**Comment by tushar-rishav** [comment]
@AbdealiJK Your blessings too :P

**Comment by SanketDG** [comment]
@tushar-rishav You will need to add a check for nltk in the `check_prerequisites` function.

File: bears/vcs/git/GitCommitBear.py [hunk 30]
```diff
@@ -23,7 +25,8 @@ def run(self,
             force_body: bool=False,
             allow_empty_commit_message: bool=False,
             shortlog_regex: str="",
-            shortlog_trailing_period: bool=None):
+            shortlog_trailing_period: bool=None,
+            shortlog_imperative_check: bool=True):
```
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: Don't you need to add this to the docstring for documentation ?

File: requirements.txt [hunk 31]
```diff
@@ -23,3 +23,6 @@ pydocstyle~=1.0.0
 cmakelint~=1.3.4
 requests_mock~=0.7.0
 vim-vint~=0.3.6
+# Stable version 3.2 is problematic in Windows.
+# Use >=3.3 once nltk-3.3 is released. 
+nltk==3.1
```
Comment (AbdealiJK) [inline requirements.txt]: Can we use ~=3.1.0 ?
They may have patched versions of 3.1 which go like 3.1.x later ?

File: tests/vcs/git/GitCommitBearTest.py (hunk 28 above)
Comment (AbdealiJK) [inline tests/vcs/git/GitCommitBearTest.py]: > IIUC "long" is actually a verb. A verb may not always be something that shows action. It could also be a description. :P

Your use of long seems to be an adjective here (from my minimal understanding of grammar).
An adjective describes a noun. A verb described an action, state, or occurrence. Here you're describing the shortlog which is a [noun](http://www.merriam-webster.com/dictionary/log) ?

[Long as a verb](http://www.oxfordlearnersdictionaries.com/definition/english/long_3): "to want something very much especially if it does not seem likely to happen soon"
Also see: [Long as an adjective](http://www.oxfordlearnersdictionaries.com/definition/english/long_1?q=long)

File: bears/vcs/git/GitCommitBear.py [hunk 32]
```diff
@@ -110,6 +115,45 @@ def check_shortlog(self,
...
 
+        if shortlog_imperative_check:
+            colon_pos = shortlog.find(':')
+            shortlog = shortlog[colon_pos + 1:] if colon_pos != -1 else shortlog
+            has_flaws = self.check_imperative(shortlog)
+            if has_flaws:
+                bad_words = ', '.join(map(itemgetter(0), has_flaws))
+                yield Result(self,
+                             "Shortlog of HEAD commit isn't imperative mood,"
+                             "bad words are '{}'".format(bad_words))
```
Comment (AbdealiJK) [inline bears/vcs/git/GitCommitBear.py]: space after comma (even in tests)

File: requirements.txt (hunk 31 above)
Comment (tushar-rishav) [inline requirements.txt]: 3.1 currently worked perfectly on Windows. Any changes would require a test
again. I'll update and see how Appveyor responds..

**Comment by tushar-rishav** [comment]
But nltk would automatically get installed as dependency. Do we still need
to check?

**Comment by sils1297** [comment]
I like it, as per discussion error rate is too high and it needs a rebase

**Comment by tushar-rishav** [comment]
@sils1297 @Makman2 @AbdealiJK  @SanketDG  as per the discussion, the error rate has dropped to 1% now. (I ran a test on coala repo commit history). Shall we proceed ahead?

**Comment by sils1297** [comment]
yes please

**Comment by sils1297** [comment]
yes please! :)

File: bears/vcs/git/GitCommitBear.py [hunk 33]
```diff
@@ -110,6 +118,45 @@ def check_shortlog(self,
...
+            A list of tuples having 2 elements (invalid word, parts of speech)
+            or an empty list if no invalid words are found.
+        """
+        try:
+            words = nltk.word_tokenize(nltk.sent_tokenize(paragraph)[0])
+            flaws = []
+            # VBZ : Verb, 3rd person singular present, like 'adds', 'writes' etc
+            # VBD : Verb, Past tense , like 'added', 'wrote' etc
+            # VBG : Verb, Present participle, like 'adding', 'writing'
+            for word, tag in nltk.pos_tag(['I'] + words)[1:2]:
```
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: no need for a loop

File: bears/vcs/git/GitCommitBear.py [hunk 34]
```diff
@@ -110,6 +118,45 @@ def check_shortlog(self,
...
+                if(tag.startswith('VBZ') or
+                   tag.startswith('VBD') or
+                   tag.startswith('VBG') or
+                   word.endswith('ing')):  # Handle special case for VBG
+                    flaws.append((word, tag))
+            return flaws
+        except LookupError as error:    # pragma: no cover
```
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: a space before # too much

File: bears/vcs/git/GitCommitBear.py [hunk 35]
```diff
@@ -110,6 +118,45 @@ def check_shortlog(self,
...
+
+        :param paragraph:
+            The input paragraph to be tested.
+        :returns:
+            A list of tuples having 2 elements (invalid word, parts of speech)
+            or an empty list if no invalid words are found.
+        """
+        try:
+            words = nltk.word_tokenize(nltk.sent_tokenize(paragraph)[0])
+            flaws = []
```
Comment (sils1297) [inline bears/vcs/git/GitCommitBear.py]: no list needed in this design right?

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala-bears/pull/315/commits/a81f4777f0679456ea1bfbc6fd0834119dff615d needs work

File: bears/vcs/git/GitCommitBear.py (hunk 35 above)
Comment (tushar-rishav) [inline bears/vcs/git/GitCommitBear.py]: right :D

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/315/commits/d888800b1b43a94291865b66a8b6a1735854a735 , can you make an ASCIINEMA?

**Comment by tushar-rishav** [comment]
@sils1297  Positively. On it :+1:

**Comment by sils1297** [comment]
@rultor merge
